allennlp/nn/util.py

```diff
@@ -162,12 +162,21 @@ def masked_log_softmax(vector, mask):
 
     We assume that both ``vector`` and ``mask`` (if given) have shape ``(batch_size, vector_dim)``.
 
-    In the case that the input vector is completely masked, this function returns an array
-    of ``0.0``.  You should be masking the result of whatever computation comes out of this in that
-    case, anyway, so it shouldn't matter.
+    In the case that the input vector is completely masked, the return value of this function is
+    arbitrary, but not ``nan``.  You should be masking the result of whatever computation comes out
+    of this in that case, anyway, so the specific values returned shouldn't matter.
+
+    If your logits are all extremely negative (i.e., the max value in your logit vector is -50 or
+    lower), the way we handle masking here could mess you up.  But if you've got logit values that
+    extreme, you've got bigger problems than this.
     """
     if mask is not None:
-        vector = vector + mask.log()
+        # vector + mask.log() is an easy way to zero out masked elements in logspace, but it
+        # results in nans when the whole vector is masked.  We need a very small value instead of a
+        # zero in the mask for these cases.  log(1 + 1e-45) is still basically 0, so we can safely
+        # just add 1e-45 before calling mask.log().  We use 1e-45 because 1e-46 is so small it
+        # becomes 0 - this is just the smallest value we can actually use.
+        vector = vector + (mask + 1e-45).log()
```

Comment:
If we pre-multiplied by the mask here, wouldn't we get more consistent behavior in the all masked, extremely negative case? It would be nice in terms of consistency if torch.exp(masked_log_softmax(all_masked_logits)) == masked_softmax(all_masked_logits), but I guess this is ok. Also, you might want to add a warning about using this function particularly with half precision floats, given that they are available in pytorch.

Reply:
`masked_softmax(all_masked_logits)` returns `zeros()`.  There's no way to make `masked_log_softmax(all_masked_logits).exp()` give the same thing, because `masked_log_softmax()` would have to return `[float('-inf'), ...]` in that case, and we can't do that without a whole lot of mostly-wasted computation.  We _could_ try to change `masked_softmax` to return something else in the all-masked case, but I don't think that's a good idea.

Are half-precision floats a global torch flag?  Like, calling `mask.float() + 1e-45` wouldn't solve the problem, right?

Reply:
Ok, agreed. No, there are explicit `torch.HalfTensor` types, but honestly allennlp would clobber most uses of this in a lot of places, so don't worry about it.